packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/mongo/MongoCollectionTests.kt

```diff
@@ -0,0 +1,1466 @@
+        assertFailsWithMessage<ServiceException>("insert not permitted") {
+            collection.insertMany(listOf(BsonDocument()))
+        }
+    }
+
+    @Test
+    fun insertMany_throwsOnTypeMismatch() = runBlocking<Unit> {
+        assertFailsWithMessage<ServiceException>("insert not permitted") {
+            collection.insertMany(listOf(BsonDocument(mapOf("_id" to ObjectId(), "name" to BsonString("object-1")))))
+        }
+    }
+
+    @Test
+    fun deleteOne() = runBlocking {
+        assertFalse { collection.deleteOne(BsonDocument()) }
+
+        assertEquals(
+            2,
+            collection.insertMany(
+                listOf(
+                    CollectionDataType("object-1"),
+                    CollectionDataType("object-1")
+                )
+            ).size
+        )
+        assertEquals(2, collection.count(BsonDocument("""{ "name": "object-1" }""")))
+
+        assertTrue { collection.deleteOne(BsonDocument("""{ "name": "object-1" }""")) }
+        assertEquals(1, collection.count(BsonDocument("""{ "name": "object-1" }""")))
+    }
+
+    @Test
+    fun deleteOne_fails() = runBlocking<Unit> {
+        assertFailsWithMessage<ServiceException>("unknown top level operator: \$who.") {
+            collection.deleteOne(BsonDocument("\$who", 1))
+        }
+    }
+
+    @Test
+    fun deleteMany() = runBlocking {
+        assertEquals(0, collection.deleteMany(BsonDocument()))
+
+        collection.insertMany((1..10).map { CollectionDataType("object-${it % 5}") })
+        assertEquals(10, collection.find().size)
+
+        assertEquals(2, collection.deleteMany(BsonDocument("""{ "name": "object-1" }""")))
+
+        assertEquals(8, collection.find().size)
+
+        assertEquals(8, collection.deleteMany(BsonDocument()))
+    }
+
+    @Test
+    fun deleteMany_fails() = runBlocking<Unit> {
+        assertFailsWithMessage<ServiceException>("unknown top level operator: \$who.") {
+            collection.deleteMany(BsonDocument("\$who", 1))
+        }
+    }
+
+    @Test
+    open fun updateOne() = runBlocking<Unit> {
+        assertEquals(0, collection.count())
+
+        assertEquals(
+            4,
+            collection.insertMany(
+                listOf(
+                    CollectionDataType("object-1"),
+                    CollectionDataType("object-1"),
+                    CollectionDataType("object-1"),
+                    CollectionDataType("object-1")
+                )
+            ).size
+        )
+        assertEquals(4, collection.count())
+
+        // Update no match
+        collection.updateOne(
+            BsonDocument("""{ "name": "NOMATCH"}"""),
+            BsonDocument("\$set", BsonDocument("""{ "name": "UPDATED"}""")),
+        ).let { (updated, upsertedId) ->
+            assertFalse(updated)
+            assertNull(upsertedId)
+        }
+
+        // Update with match match
+        collection.updateOne(
+            BsonDocument("""{ "name": "object-1"}"""),
```

Comment:
We could use named parameters to improve readability.

Reply:
Yeah, it can be a bit difficult to decode with all these BsonDocuments. I have tried adding argument names, where I forgot.